jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java

```diff
@@ -0,0 +1,306 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.Query;
+
+import org.jbpm.kie.services.impl.model.ProcessInstanceWithVarsDesc;
+import org.jbpm.kie.services.impl.model.UserTaskInstanceWithPotOwnerDesc;
+import org.jbpm.shared.services.impl.QueryManager;
+import org.jbpm.shared.services.impl.TransactionalCommandService;
+import org.jbpm.shared.services.impl.commands.QueryNameCommand;
+import org.kie.api.runtime.query.QueryContext;
+
+import static java.util.Collections.singletonMap;
+
+public abstract class AbstractAdvanceRuntimeDataServiceImpl {
+
+    private EntityManagerFactory emf;
+    private TransactionalCommandService commandService;
+
+    public AbstractAdvanceRuntimeDataServiceImpl() {
+        QueryManager.get().addNamedQueries("META-INF/Servicesorm.xml");
+    }
+    public void setCommandService(TransactionalCommandService commandService) {
+        this.commandService = commandService;
+    }
+
+    public void setEmf(EntityManagerFactory emf) {
+        this.emf = emf;
+    }
+
+
+    public List<org.jbpm.services.api.model.ProcessInstanceWithVarsDesc> queryProcessByVariables(Map<String, Object> attributes,
+                                                                                                 Map<String, Object> variables,
+                                                                                                 int processType,
+                                                                                                 String varPrefix,
+                                                                                                 QueryContext queryContext) {
+        EntityManager entityManager = emf.createEntityManager();
+
+        // first step is to filter the data creating a derived tables and pivoting var - rows to columns (only the variables we are interested to filter)
+        StringBuilder where = new StringBuilder();
+        StringBuilder derivedTables = new StringBuilder();
+        for (Map.Entry<String, Object> var : variables.entrySet()) {
+            String alias = "TABLE_" + var.getKey();
+            String varName = var.getKey();
+
+            derivedTables.append("LEFT JOIN (" +
+                                 "SELECT A1.processId, A1.processInstanceId, A1.variableId AS NAME, A1.value AS VALUE \n" +
+                                 "FROM VariableInstanceLog A1 \n" +
+                                 "LEFT JOIN VariableInstanceLog A2 ON A1.processId = A2.processId AND A1.processInstanceId = A2.processInstanceId AND A1.variableInstanceId = A2.variableInstanceId AND A2.log_date < A1.log_date  \n" +
+                                 "WHERE A2.log_date IS NULL \n" +
+                                 ") " + alias + " ON " + alias + ".processId = pil.processId AND " + alias + ".processInstanceId = pil.processInstanceId \n");
```

Comment:
Do not use log_date, but id, for ordering, as log_date may be the same.
Invert the order to pick up the current value